File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/ser/binary/types/PSerializer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.tinkerpop.gremlin.driver.ser.binary.types;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.ByteBufAllocator;
import io.netty.buffer.CompositeByteBuf;
import org.apache.tinkerpop.gremlin.driver.ser.SerializationException;
import org.apache.tinkerpop.gremlin.driver.ser.binary.DataType;
import org.apache.tinkerpop.gremlin.driver.ser.binary.GraphBinaryReader;
import org.apache.tinkerpop.gremlin.driver.ser.binary.GraphBinaryWriter;
import org.apache.tinkerpop.gremlin.process.traversal.P;
import org.apache.tinkerpop.gremlin.process.traversal.util.AndP;
import org.apache.tinkerpop.gremlin.process.traversal.util.ConnectiveP;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

/**
 * @author Stephen Mallette (http://stephen.genoprime.com)
 */
public class PSerializer<T extends P> extends SimpleTypeSerializer<T> {

    private final Class<T> classOfP;

    public PSerializer(final DataType typeOfP, final Class<T> classOfP) {
        super(typeOfP);
        this.classOfP = classOfP;
    }

    @Override
    T readValue(final ByteBuf buffer, final GraphBinaryReader context) throws SerializationException {
        final String predicateName = context.readValue(buffer, String.class, false);
        final int length = context.readValue(buffer, Integer.class, false);
        final Object[] args = new Object[length];
        final Class<?>[] argumentClasses = new Class[length];
        boolean collectionType = false;
        for (int i = 0; i < length; i++) {
            args[i] = context.read(buffer);
            argumentClasses[i] = args[i].getClass();
        }

        switch (predicateName) {
            case "and":
                return (T) ((P) args[0]).and((P) args[1]);
            case "or":
                return (T) ((P) args[0]).or((P) args[1]);
            default:
                Method m;
                try {
                    // do a direct lookup
                    m = classOfP.getMethod(predicateName, argumentClasses);
                } catch (NoSuchMethodException ex0) {
                    // then try collection types
                    try {
                        m = classOfP.getMethod(predicateName, Collection.class);
                        collectionType = true;
                    } catch (NoSuchMethodException ex1) {
                        // finally go for the generics
                        try {
                            m = classOfP.getMethod(predicateName, Object.class);
                        } catch (NoSuchMethodException ex2) {
                            throw new SerializationException("not found");
                        }
                    }
                }

                try {
                    if (Modifier.isStatic(m.getModifiers())) {
                        if (collectionType)
                            return (T) m.invoke(null, Arrays.asList(args));
                        else
                            return (T) m.invoke(null, args);
                    } else {
                        // try an instance method as it might be a form of ConnectiveP which means there is a P as an
                        // argument that should be used as the object of an instance method
                        if (args.length != 2 || !(args[0] instanceof P) || !(args[1] instanceof P))
                            throw new IllegalStateException(String.format("Could not determine the form of P for %s and %s",
                                    predicateName, Arrays.asList(args)));

                        final P firstP = (P) args[0];
                        final P secondP = (P) args[1];

                        return (T) m.invoke(firstP, secondP);
                    }
                } catch (Exception ex) {
                    throw new SerializationException(ex);
                }
        }
    }

    @Override
    public ByteBuf writeValue(final T value, final ByteBufAllocator allocator, final GraphBinaryWriter context) throws SerializationException {
        // the predicate name is either a static method of P or an instance method when a type ConnectiveP
        final boolean isConnectedP = value instanceof ConnectiveP;
        final String predicateName = isConnectedP ?
                (value instanceof AndP ? "and" : "or") :
                value.getBiPredicate().toString();
        final Object args = isConnectedP ? ((ConnectiveP<?>) value).getPredicates() : value.getValue();

        final List<Object> argsAsList = args instanceof Collection ? new ArrayList<>((Collection) args) : Collections.singletonList(args);
        final int length = argsAsList.size();
        final CompositeByteBuf result = allocator.compositeBuffer(2 + length);

        result.addComponent(true, context.writeValue(predicateName, allocator, false));
        result.addComponent(true, context.writeValue(length, allocator, false));

        for (Object o : argsAsList) {
            result.addComponent(true, context.write(o, allocator));
        }

        return result;
    }
}

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/io/binary/internals/PSerializer.js
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

/**
 * @author Igor Ostapenko
 */
'use strict';

const t = require('../../../../process/traversal');

module.exports = class PSerializer {
  constructor(ioc) {
    this.ioc = ioc;
    this.ioc.serializers[ioc.DataType.P] = this;
  }

  canBeUsedFor(value) {
    return value instanceof t.P;
  }

  serialize(item, fullyQualifiedFormat = true) {
    if (item === undefined || item === null) {
      if (fullyQualifiedFormat) {
        return Buffer.from([this.ioc.DataType.P, 0x01]);
      }
      const name = [0x00, 0x00, 0x00, 0x00]; // ''
      const values_length = [0x00, 0x00, 0x00, 0x00]; // 0
      return Buffer.from([...name, ...values_length]);
    }

    const bufs = [];
    if (fullyQualifiedFormat) {
      bufs.push(Buffer.from([this.ioc.DataType.P, 0x00]));
    }

    // {name}
    bufs.push(this.ioc.stringSerializer.serialize(item.operator, false));

    // {values_length}{value_0}...{value_n}
    //
    // This is a tricky part due to the nature of t.P.
    // GraphBinary states that {name} is followed by list only,
    // but t.P may be constructed such way that its item.value can be
    // a usual string or something, that's why we should care here
    // to get it wrapped into a list.
    //
    // It tries to resemble the same if-else structure as
    // GraphSON.PSerializer.serialize() does.
    let list;
    if (item.other === undefined || item.other === null) {
      if (Array.isArray(item.value)) {
        list = item.value;
      } else {
        list = [item.value];
      }
    } else {
      list = [item.value, item.other];
    }
    bufs.push(this.ioc.listSerializer.serialize(list, false));

    return Buffer.concat(bufs);
  }

  deserialize(buffer, fullyQualifiedFormat = true) {
    let len = 0;
    let cursor = buffer;

    try {
      if (buffer === undefined || buffer === null || !(buffer instanceof Buffer)) {
        throw new Error('buffer is missing');
      }
      if (buffer.length < 1) {
        throw new Error('buffer is empty');
      }

      if (fullyQualifiedFormat) {
        const type_code = cursor.readUInt8();
        len++;
        if (type_code !== this.ioc.DataType.P) {
          throw new Error('unexpected {type_code}');
        }
        cursor = cursor.slice(1);

        if (cursor.length < 1) {
          throw new Error('{value_flag} is missing');
        }
        const value_flag = cursor.readUInt8();
        len++;
        if (value_flag === 1) {
          return { v: null, len };
        }
        if (value_flag !== 0) {
          throw new Error('unexpected {value_flag}');
        }
        cursor = cursor.slice(1);
      }

      let name, name_len;
      try {
        ({ v: name, len: name_len } = this.ioc.stringSerializer.deserialize(cursor, false));
        len += name_len;
      } catch (err) {
        err.message = '{name}: ' + err.message;
        throw err;
      }
      cursor = cursor.slice(name_len);

      let values, values_len;
      try {
        ({ v: values, len: values_len } = this.ioc.listSerializer.deserialize(cursor, false));
        len += values_len;
      } catch (err) {
        err.message = '{values}: ' + err.message;
        throw err;
      }
      cursor = cursor.slice(values_len);

      if (values.length < 1) {
        return { v: new t.P(''), len };
      }

      let v;
      const P_static = t.P[name];
      if (typeof P_static === 'function') {
        v = P_static(...values); // it's better to follow existing logic which may depend on an operator name
      } else {
        v = new t.P(name, ...values);
      }

      return { v, len };
    } catch (err) {
      throw this.ioc.utils.des_error({ serializer: this, args: arguments, cursor, err });
    }
  }
};
